File: Planet.java
/*
Objective: for the structure of a planet node in the network with its needs, gives, and name
Author: Nathan Chapman
Date: 9/8/22
*/

import java.util.ArrayList;

public class Planet {

  public static String[] resources  = {"food", "water", "air", "materials", "people", "entertainment"};
  public static String needsGivesState = "NG", needsState = "N_", givesState = "_G", satisfiedState = "__";

  // Instrinsic properties
  public String name;
  private int id;
  // values of strings can be:
  // "NG" = has unmet needs, and has things to give (default state)
  // "N_" = has unmet needs, but can't give any more
  // "_G" = has needs met, but still has resources to give
  // "__" = has needs met and no more resources to give
  public String   state = needsGivesState;
  public ArrayList<String> needs = new ArrayList<String>(2);
  public ArrayList<String> gives = new ArrayList<String>(2);

  public Planet(int count, int num, ArrayList<String> needs, ArrayList<String> gives) {
    this.id    = count;
    this.name  = Integer.toHexString(num).toUpperCase();
    this.needs = needs;
    this.gives = gives;
  }

  public String toString () {
    String nameString  = "ID:    "   + name;
    String stateString = "\nState: " + state;
    String needsString = "\nNeeds: " + needs.get(0) +", " + needs.get(1);
    String givesString = "\nGives: " + gives.get(0) + ", " + gives.get(1);
    String connectionsString = "\nConnected to: " + "CONNCETIONS";
    return  nameString + needsString + givesString + "\n";
  }
}

File: python/planet.py
# planet class

import random as rng

RESOURCE_LIST  = ["food", "water", "air", "materials", "people", "entertainment"]
NEEDS_GIVES_STATE = "NG"
NEEDS_STATE = "N_"
GIVES_STATE = "_G"
SATISFIED_STATE = "__"

class Planet:
    # TODO: add static planet count and use it as the id
    def __init__(self):
        self.name = hex(rng.randint(16**6, 16**7 - 1))[2:].upper() # make the name a 6 character hex string
        self.id   = "Come back to this"
        self.state = NEEDS_GIVES_STATE
        self.needs = {*rng.sample(RESOURCE_LIST, 2)}
        self.gives = {*rng.sample(RESOURCE_LIST, 2)}

    def summary(self):
        print(f"name: {self.name}")
        print(f"id: {self.id}")
        print(f"state: {self.state}")
        print(f"needs: {self.needs if self.needs != {} else 'needs met!'}")
        print(f"gives: {self.gives if self.gives != {} else 'gives met!'}")

    def update_state(self):
        if self.needs == {} and self.gives == {}:
            self.state = SATISFIED_STATE
        elif self.needs == {}:
            self.state = GIVES_STATE
        elif self.gives == {}:
            self.state == NEEDS_STATE
        else:
            self.state = NEEDS_GIVES_STATE

def update_needs_gives(planet_1, planet_2):
    needs_gives_intersection = planet_1.needs.intersection(planet_2.gives)
    for resource in needs_gives_intersection:
        planet_1.needs.remove(resource)
        planet_2.gives.remove(resource)

def test():
    p1 = Planet()
    p1.summary()